wip-mariocao-min-difficulty-dr: draft submission (#58)
witnet/WIPs: wip-mariocao-min-difficulty-dr.md

- `rf` is the data request replication factor

The probability `P` will be used as a static threshold that sets the maximum value until which a _RepPoE_ proof can be considered valid. This threshold will be compared with the output of a [VRF] of the identity which takes as seed the data request hash and the last checkpoint beacon of the chain (i.e. epoch and block hash).

Later, a factor `alpha` was introduced to adjust the data request mining eligibilities by taking into consideration the state of the reputation system. This factor aims to compensate for the over-representation of identities with significant reputation scores and thus affecting the average amount of eligible identities per data request.

```
P_i = reputation_i / R * rf * alpha
```

where:

- `alpha` is a factor as defined in [Dynamic Threshold]

The last modification to the _RepPoE_ algorithm was the usage of a metric labeled as _identity eligibility_ instead of the reputation score as the numerator of the previous formula:

```
P_i = eligibility_i / R * rf * alpha
```

where:

- `eligibility_i` is the computed identity eligibility as mentioned in [witnet-rust#1496] and later described

> The _eligibility_ metric aims to achieve a "more linear" distribution of data request mining probabilities across identities. As shown in the images below, the eligibility can be understood as a redistribution of reputation across identities following a _right trapezoid_ shape.

Author: > What's the subtext behind the quotation marks?

First, it was to remind me to review that adjective.
I wanted to avoid using "fairer distribution", which can be considered a subjective qualifier.

Anyhow, I forgot to remove the double quotes.

Fig. 2 - Application of the <i>eligibility algorithm</i> in previous sorted identities.

<br/>

The _eligibility_ algorithm can be decomposed into 5 simple steps:

1. Identities are sorted by reputation in descending order (see Fig.1).

2. The upper edge of the trapezoid (the triangle) is calculated from the line following the equation `y = m*x + k`, where:

    ```rust
    let k = 1.5 * (average_reputation - minimum_reputation);
    let m = -k / (active_reputed_identities_length - 1);
    ```

3. The curve is used to assign the initial eligibility to the previously sorted identities (see _Step 3: triangle_ in Fig.2).

4. The identities eligibility is offset by adding the remaining reputation points that were not redistributed in the previous step (see _Step 4: rectangle_ in Fig.2).

    ```rust
    let remaining_reputation = total_active_reputation - total_triangle_reputation;
    let offset_reputation = remaining_reputation / (active_reputed_identities_length);
    ```

    This step is important to keep the same data request mining difficulty for newcomers or Sybil identities.

> 5. The rest of the `offset_reputation` integer division is assigned (in order) to the sorted identities until depleted (see _Step 5: remaining_ in Fig.2).

Reviewer: ```suggestion
5. The remainder of the `offset_reputation` integer division is assigned (in order) to the sorted identities until depleted (see _Step 5: remaining_ in Fig.2).
```

    ```rust
    let identities_with_extra_eligibility = (remaining_reputation) % active_reputed_identities_length;
    ```

Steps 4 and 5 aim to preserve the sum of reputation equal to the sum of eligibility. Geometrically, this can be defined as matching the area of the trapezoid (Fig.2) with the area of the reputation curve (Fig.1).

The identity's reputation, from the perspective of chain state, is not altered after applying the [RepPoE] algorithm.

## Specification

### Removal of the dynamic threshold

**(1)**: The [Dynamic Threshold] factor MUST be removed from the _RepPoE_ algorithm.

As eligibility across reputed identities is fairer due to the changes described in [witnet-rust#1496], the [Dynamic Threshold] does not have anymore a significant impact on the probability of mining data requests. Additionally, the [Dynamic Threshold] was not properly updated to use the eligibility metrics instead of the reputation scores.

> In case of abnormally low ARS sizes, it is assumed that there are in the order of thousands of non-reputed identities computing their eligibility to be selected to solved the data request. This assumption is in line with the minimum difficulty set in [WIP-0012] and previous network conditions documented in [WIP-0009].

Reviewer: ```suggestion
In case of abnormally low ARS sizes, it is assumed that there is a number of non-reputed identities in the order of thousands computing their eligibility to be selected to solve the data request. This assumption is in line with the minimum difficulty set in [WIP-0012] and previous network conditions documented in [WIP-0009].
```

### Introduction of a minimum data request mining difficulty

**(2)**: A lower bound on data request mining difficulty is proposed. For all data requests following the activation of this proposal, eligibility MUST be computed using the greater of two values: the ARS size and a constant `minimum_difficulty`:

```rust
let dr_eligibity = identity_eligibility / max(total_active_reputation, minimum_difficulty) * replication_factor
```

### Equate the data request minimum difficulty to the mining difficulty

> Analogously to [WIP-0012], such lower bound to be effective, its value needs to be strictly greater than `replication_factor`, as otherwise, ARS members will have an eligibility equal or greater than 100%.

Reviewer: ```suggestion
Analogously to [WIP-0012], for such lower bound to be effective, its value needs to be strictly greater than `replication_factor`, as otherwise, ARS members will have an eligibility equal or greater than 100%.
```

> In [WIP-0012], the `initial_difficulty` consensus constant was repurposed as `minimum_difficulty` for mining blocks.

Reviewer: I think we could omit this detail that gives some color on WIP-0012 but doesn't contribute much to this specification.

**(3)**: A value of 2000 (two thousand) is set for the data request minimum difficulty so that it equals the `minimum_difficulty` defined in [WIP-0012].

## Backwards compatibility

- Implementer: a client that implements or adopts the protocol improvements proposed in this document.
- Non-implementer: a client that fails to or refuses to implement the protocol improvements proposed in this document.

#### Consensus cheat sheet

Upon entry into force of the proposed improvements:

> - Block and transactions that were considered valid by former versions of the protocol MUST continue to be considered valid.

Reviewer: ```suggestion
- Blocks and transactions that were considered valid by former versions of the protocol MUST continue to be considered valid.
```

- Blocks and transactions produced by non-implementers MUST be validated by implementers using the same rules as with those coming from implementers, that is, the new rules.
- Implementers MUST apply the proposed logic when evaluating transactions or computing their own data request eligibility.

As a result:

- Non-implementers MAY NOT get their blocks and transactions accepted by implementers.
- Implementers MAY NOT get their valid blocks accepted by non-implementers.

> - Non-implementers MAY consolidate blocks and superblocks than implementers.

Reviewer 2: Different? Other?

```suggestion
- Non-implementers MAY consolidate different blocks and superblocks than implementers.
```
